GP2 external tests failure due to unknown properties in type `HardhatUserConfig`

## Description

`hardhat-deploy` extends `HardhatUserConfig` type (see [here](https://github.com/wighawag/hardhat-deploy/blob/master/src/type-extensions.ts#L11)) by redeclaring the module where it was originally defined in [hardhat-core](https://github.com/NomicFoundation/hardhat/blob/main/packages/hardhat-core/src/types/config.ts#L274). This should work fine, and in fact has been working for a while, but apparently the gp2 tests started to fail with the following error in our CI ([example](https://app.circleci.com/pipelines/github/ethereum/solidity/28455/workflows/010fd2c3-8a72-45b1-b7d3-d9de508da73f/jobs/1263331)):

```
Running compile function...
yarn run v1.22.18
$ yarn build:sol && yarn build:ts
$ hardhat compile --force
An unexpected error occurred:

hardhat.config.ts:131:3 - error TS2322: Type '{ mocha: MochaOptions; paths: { artifacts: string; cache: string; deploy: string; sources: string; }; solidity: { compilers: ({ version: string; settings: { optimizer: { enabled: boolean; runs: number; }; }; } | { ...; })[]; }; networks: { ...; }; namedAccounts: { ...; }; gasReporter: { ...; }; }' is not assignable to type 'HardhatUserConfig'.
  Object literal may only specify known properties, and 'namedAccounts' does not exist in type 'HardhatUserConfig'.

131   namedAccounts: {
      ~~~~~~~~~~~~~~~~
132     // Note: accounts defined by a number refer to the the accounts as configured
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
...
146     },
    ~~~~~~
147   },
    ~~~
hardhat.config.ts:169:8 - error TS2339: Property 'gasReporter' does not exist on type 'HardhatUserConfig'.

169 config.gasReporter = {
           ~~~~~~~~~~~
hardhat.config.ts:185:8 - error TS2339: Property 'gasReporter' does not exist on type 'HardhatUserConfig'.

185 config.gasReporter = {
           ~~~~~~~~~~~
```

The error suggests that hardhat is attempting to use a `HardhatUserConfig` that does not contain `namedAccounts` property. But this property should have been added by the `hardhat-deploy` plugin.

After some debugging, the problem seems to be a conflict in the import of `HardhatUserConfig` when both `hardhat-waffle` and `hardhat-deploy` are also imported, like in the [gp2's hardhat.config.ts](https://github.com/cowprotocol/contracts/blob/main/hardhat.config.ts#L1). For instance, doing the following:

```javascript
import { HardhatUserConfig } from 'hardhat/types';
import "@nomiclabs/hardhat-waffle";
import "hardhat-deploy";

const config: HardhatUserConfig = { /* ... */ }
```

Resolves to config being of the type `HardhatUserConfig` redeclared by `@nomiclabs/hardhat-waffle` which overrides the module `hardhat/types`. See: https://github.com/NomicFoundation/hardhat-waffle/blob/main/src/index.ts#L15.

And doing:

```javascript
import { HardhatUserConfig } from 'hardhat/types';
import "hardhat-deploy";
import "@nomiclabs/hardhat-waffle";

const config: HardhatUserConfig = { /* ... */ }
```

Resolves  to config being of the type `HardhatUserConfig` redeclared by `hardhat-deploy` which overrides the module `hardhat/types/config`. See: https://github.com/wighawag/hardhat-deploy/blob/master/src/type-extensions.ts#L11

The same problem happens when importing  `@nomiclabs/hardhat-waffle` and `hardhat-gas-reporter` in the same file.

I suspect the problem was introduced by the recent migration of `hardhat-waffle` from hardhat repository to its [own repository](https://github.com/NomicFoundation/hardhat-waffle). This was done ~two days ago: https://github.com/NomicFoundation/hardhat/commit/dadad32dde94856e6c859a800563f9961846f685
With the migration, now `hardhat-waffle` redeclares the `hardhat/types` to extend it, which it doesn't do before the migration to the new repository (see [here](https://github.com/NomicFoundation/hardhat/commit/dadad32dde94856e6c859a800563f9961846f685#diff-4eb1c0b2b2315d5d45c833cb72f7dc48f518de5bdc4d16ac48cd563571d6ba02)).

## Workaround

We could just change the order of the imports to fix the problem, i.e. moving `hardhat-waffle` to be imported after `hardhat-deploy` and `hardhat-gas-reporter`.

Another option is to import the `HardhatUserConfig` from the `hardhat/types/config` module which is the one overridden by `hardhat-deploy` and `hardhat-gas-reporter` instead of importing it from `hardhat/types` (as our current script does) which is overridden by `hardhat-waffle`. Example:

```javascript
import { HardhatUserConfig } from "hardhat/types/config"; // Works without changing the order of imports
// import { HardhatUserConfig } from "hardhat/config"; // Conflicts
// import { HardhatUserConfig } from "hardhat/types"; // Conflicts
import "@nomiclabs/hardhat-waffle";
import "hardhat-deploy";
import "hardhat-gas-reporter";
// ...
// Moving the hardhat-waffle import to here also works using any of the `HardhatUserConfig` alias imports above
// import "@nomiclabs/hardhat-waffle";

const config: HardhatUserConfig = {
  solidity: "0.8.17",
  namedAccounts: {
    deployer: 0,
  },
  gasReporter: {
    enabled: true,
  },
};
```

--- a/test/externalTests/common.sh
+++ b/test/externalTests/common.sh
@@ -235,8 +235,7 @@ function force_truffle_compiler_settings
 function name_hardhat_default_export
 {
     local config_file="$1"
-
-    local import="import {HardhatUserConfig} from 'hardhat/types';"
+    local import="import {HardhatUserConfig} from 'hardhat/types/config';"
     local config="const config: HardhatUserConfig = {"
     sed -i "s|^\s*export\s*default\s*{|${import}\n${config}|g" "$config_file"
     echo "export default config;" >> "$config_file"

--- a/test/externalTests/gp2.sh
+++ b/test/externalTests/gp2.sh
@@ -76,6 +76,9 @@ function gp2_test
     # See https://github.com/cowprotocol/contracts/issues/32
     yarn add @tenderly/hardhat-tenderly@1.1.6
 
+    # Add missing sinon type definitions
+    yarn add @types/sinon
+
     # Some dependencies come with pre-built artifacts. We want to build from scratch.
     rm -r node_modules/@gnosis.pm/safe-contracts/build/
 

